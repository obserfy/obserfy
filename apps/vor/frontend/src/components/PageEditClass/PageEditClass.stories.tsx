import { FC } from "react"
import PageEditClass from "./PageEditClass"

export default {
  title: "Core/PageEditClass",
  component: PageEditClass,
  parameters: {
    componentSubtitle: "Just a simple PageEditClass",
  },
}

export const Basic: FC = () => <PageEditClass classId="asdfasdfa" />
