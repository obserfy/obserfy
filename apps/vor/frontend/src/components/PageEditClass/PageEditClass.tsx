import { FC, useEffect, useRef, useState } from "react"
import { useImmer } from "use-immer"
import { Flex, Button, Box } from "theme-ui"
import { Trans } from "@lingui/macro"
import { navigate } from "../Link/Link"
import { CLASS_SETTINGS_URL } from "../../routes"
import BackNavigation from "../BackNavigation/BackNavigation"
import { Typography } from "../Typography/Typography"
import Input from "../Input/Input"
import Chip from "../Chip/Chip"
import { WEEKDAYS } from "../PageNewClass/PageNewClass"
import useGetClass from "../../hooks/api/classes/useGetClass"
import LoadingPlaceholder from "../LoadingPlaceholder/LoadingPlaceholder"
import usePatchClass from "../../hooks/api/classes/usePatchClass"
import LoadingIndicator from "../LoadingIndicator/LoadingIndicator"
import DeleteClassDialog from "../DeleteClassDialog/DeleteClassDialog"
import ErrorMessage from "../ErrorMessage/ErrorMessage"
import dayjs from "../../dayjs"

interface Props {
  classId: string
}
export const PageEditClass: FC<Props> = ({ classId }) => {
  const [name, setName] = useState("")
  const [startTime, setStartTime] = useState("09:00")
  const [endTime, setEndTime] = useState("10:00")
  const [weekdays, setWeekdays] = useImmer<number[]>([])
  const [showDeleteDialog, setShowDeleteDialog] = useState(false)
  const { mutateAsync, status, error } = usePatchClass(classId)
  const classes = useGetClass(classId)
  const isLoaded = useRef(false)
  const valid = name !== ""

  useEffect(() => {
    const target = classes.data
    if (!target || isLoaded.current) return
    const start = Date.parse(target.startTime)
    const end = Date.parse(target.endTime)
    setName(target.name)
    setStartTime(dayjs(start).format("HH:mm"))
    setEndTime(dayjs(end).format("HH:mm"))
    setWeekdays(() => target.weekdays)
    isLoaded.current = true
  }, [classes.data, setWeekdays])

  const patchClass = async (): Promise<void> => {
    try {
      await mutateAsync({
        name,
        weekdays,
        endTime: dayjs(endTime, "HH:mm").toDate(),
        startTime: dayjs(startTime, "HH:mm").toDate(),
      })
      await navigate(CLASS_SETTINGS_URL)
    } catch (e) {
      Sentry.captureException(e)
    }
  }

  return (
    <>
      <Box sx={{ maxWidth: "maxWidth.sm" }} margin="auto">
        <BackNavigation to={CLASS_SETTINGS_URL} text="Class" />
        <Typography.H5 m={3}>
          <Trans>Edit Class</Trans>
        </Typography.H5>
        {classes.status === "loading" && <LoadingState />}
        {classes.status === "success" && (
          <Box p={3}>
            <Input
              label="Name"
              sx={{ width: "100%" }}
              mb={3}
              value={name}
              onChange={(e) => setName(e.target.value)}
            />
            <Flex>
              <Input
                type="time"
                value={startTime}
                onChange={(e) => setStartTime(e.target.value)}
                label="Start Time"
                sx={{ width: "100%" }}
                mb={3}
                mr={3}
              />
              <Input
                type="time"
                value={endTime}
                onChange={(e) => setEndTime(e.target.value)}
                label="End Time"
                sx={{ width: "100%" }}
                mb={3}
              />
            </Flex>
            <Typography.Body mb={2} color="textMediumEmphasis">
              <Trans>Available every</Trans>
            </Typography.Body>
            <Flex
              mb={3}
              sx={{
                flexWrap: "wrap",
              }}
            >
              {WEEKDAYS.map((weekday, i) => (
                <Chip
                  mb={2}
                  mr={2}
                  key={weekday}
                  text={weekday}
                  activeBackground="primary"
                  isActive={weekdays.includes(i)}
                  onClick={() =>
                    setWeekdays((draft) => {
                      if (draft.includes(i)) {
                        return draft.filter((item) => item !== i)
                      }
                      return [...draft, i]
                    })
                  }
                />
              ))}
            </Flex>
            <Flex>
              <Button
                variant="outline"
                mr={3}
                color="danger"
                sx={{ flexShrink: 0 }}
                onClick={() => setShowDeleteDialog(true)}
              >
                <Trans>Delete</Trans>
              </Button>
              <Button
                sx={{ flexGrow: 1 }}
                disabled={!valid || status === "loading"}
                onClick={patchClass}
              >
                {status === "loading" && (
                  <LoadingIndicator mr={2} color="onPrimary" />
                )}
                <Trans>Save</Trans>
              </Button>
            </Flex>
            <ErrorMessage error={error} />
          </Box>
        )}
      </Box>
      {showDeleteDialog && (
        <DeleteClassDialog
          onDismiss={() => setShowDeleteDialog(false)}
          classId={classId}
          name={classes.data?.name ?? ""}
        />
      )}
    </>
  )
}

const LoadingState: FC = () => {
  return (
    <Box m={3} mt={4}>
      <LoadingPlaceholder sx={{ width: "12rem", height: 48 }} mb={3} />
      <LoadingPlaceholder sx={{ width: "100%", height: 48 }} mb={3} />
      <Flex mb={3}>
        <LoadingPlaceholder sx={{ width: "100%", height: 48 }} mb={3} />
        <LoadingPlaceholder sx={{ width: "100%", height: 48 }} />
      </Flex>
      <LoadingPlaceholder sx={{ width: "100%", height: 48 }} mb={3} />
      <LoadingPlaceholder sx={{ width: "100%", height: 48 }} mb={3} />
    </Box>
  )
}

export default PageEditClass
