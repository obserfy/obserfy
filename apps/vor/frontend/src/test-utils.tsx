/* eslint-disable import/no-extraneous-dependencies */
import { FC, ReactElement } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { ThemeProvider, Box, Theme } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import { I18nProvider } from "@lingui/react"
import { i18n } from "@lingui/core"
import { en } from "make-plural/plurals"
import theme from "./gatsby-plugin-theme-ui"
// @ts-ignore
import enCatalog from "../i18n/lingui/en/messages.js"

// any used here because the catalogs are generated b lingui
// not our responsibility
i18n.loadLocaleData("en", { plurals: en })
i18n.load("en", enCatalog.messages as any)
i18n.activate("en")

const queryClient = new QueryClient()
const history = createHistory(createMemorySource("/"))

const AllProviders: FC = ({ children }) => {
  return (
    <I18nProvider i18n={i18n}>
      <QueryClientProvider client={queryClient}>
        <LocationProvider history={history}>
          <ThemeProvider theme={theme as Theme}>
            <Box sx={{ fontSize: [16, 20] }}>{children}</Box>
          </ThemeProvider>
        </LocationProvider>
      </QueryClientProvider>
    </I18nProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
