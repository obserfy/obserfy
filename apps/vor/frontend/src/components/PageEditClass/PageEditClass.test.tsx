import { render } from "../../test-utils"
import { Basic } from "./PageEditClass.stories"

describe("PageEditClass", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
