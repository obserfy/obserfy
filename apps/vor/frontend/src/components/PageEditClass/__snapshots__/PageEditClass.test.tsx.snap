// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageEditClass should render correctly 1`] = `
@keyframes animation-0 {
  0% {
    opacity: 0.05;
  }

  50% {
    opacity: 0.2;
  }

  100% {
    opacity: 0.05;
  }
}

@keyframes animation-0 {
  0% {
    opacity: 0.05;
  }

  50% {
    opacity: 0.2;
  }

  100% {
    opacity: 0.05;
  }
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width: 40em) {
  .emotion-0 {
    font-size: 20px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin: auto;
  max-width: 640px;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: -8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  margin: 16px;
  margin-right: 4px;
  fill: var(--theme-ui-colors-icon);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  fill: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: var(--theme-ui-colors-text);
  font-size: 14px;
  text-rendering: optimizeLegibility;
  line-height: 1.5;
  margin-bottom: 0;
  color: var(--theme-ui-colors-textMediumEmphasis);
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: bold;
  line-height: heading;
  color: var(--theme-ui-colors-text);
  text-rendering: optimizeLegibility;
  font-size: 24px;
  margin: 16px;
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin: 16px;
  margin-top: 32px;
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 16px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  width: 12rem;
  height: 48px;
}

.emotion-8 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 16px;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  width: 100%;
  height: 48px;
}

.emotion-9 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  border-radius: 6px;
  -webkit-animation: 4s infinite animation-0;
  animation: 4s infinite animation-0;
  background-color: #a5a5a5;
  width: 100%;
  height: 48px;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <a
        href="/dashboard/admin/class"
        style="display: block;"
      >
        <div
          class="emotion-2"
        >
          <div
            class="emotion-3"
          />
          <p
            class="emotion-4"
          >
            Class
          </p>
        </div>
      </a>
      <h5
        class="emotion-5"
      >
        Edit Class
      </h5>
      <div
        class="emotion-6"
      >
        <div
          class="emotion-7"
        />
        <div
          class="emotion-8"
        />
        <div
          class="emotion-9"
        >
          <div
            class="emotion-8"
          />
          <div
            class="emotion-11"
          />
        </div>
        <div
          class="emotion-8"
        />
        <div
          class="emotion-8"
        />
      </div>
    </div>
  </div>
</div>
`;
